zerver/lib/import_realm.py

```diff
@@ -1470,6 +1470,8 @@ def import_message_data(realm: Realm, sender_map: Dict[int, Record], import_dir:
         message_id_map = ID_MAP["message"]
         for row in data["zerver_message"]:
             row["id"] = message_id_map[row["id"]]
+            del row["realm"]
```

Comment:
Why is this necessary? I would expect the slack/gitter/etc. export tools to not have a `realm` field here, and it makes me wonder whether we should either be calling `re_map_foreign_keys(data, 'zerver_message', 'realm')` above, and/or doing something different in the export tool.

Reply:
Hmm yeah `re_map_foreign_keys` might be more natural here and it does the same thing (I wasn't sure if `realm` wasn't a special case that wouldn't be handled correctly by `re_map` but it seems it's fine)

Reply:
Hmm, actually no, this breaks Slack/Gitter/etc. tests because `re_map` ends up setting `realm_id` to `None` - because the `realm` on these `data["zerver_message"]` rows is `None`. Which makes sense, since they comes from other tools, so they don't have a realm.

I'm not sure if there's a better approach, but I think given this, it seems natural to just explicitly set `row["realm_id"]` to `realm.id` here rather than trying to go through helper functions that aren't quite made for this?

Reply:
Which specific tests? I believe with these changes to the import/export tools, we should be setting `realm_id` in these data import tools, right? I would have thought this line would ensure `realm_id` was never `None` on messages exported from these third-party tools:

https://github.com/zulip/zulip/pull/23084/files#diff-0813bd0056bf62f96c31c6d0dacb4f12f68fd10c813dc923ab44e5a591817502

I don't think there should be anything special about this particular code path (but there might be some tests that do something hacky?).

Reply:
Hmm, currently the `realm_id` involved here is a dummy value, set to `0`:
```
def do_convert_data(original_path: str, output_dir: str, token: str, threads: int = 6) -> None:
    # Subdomain is set by the user while running the import command
    realm_subdomain = ""
    realm_id = 0
```
not sure why the dummy value is plumbed through to be fair. Gitter is the same, sets `realm_id = 0` at the top level. Ultimately, `realm_id` cannot be known at the time of being processed in the data conversion codepaths, since that only gets "discovered" upon creating the `Realm` in `import_realm.py` already. So this being `None` in the converted Messages and needing to get explicitly set makes sense to me :thinking:

Reply:
I see; let's discuss in a #backend thread. It sounds like we don't handle `realm_id` the way we handle all the other foreign keys, and that's probably a bug of sorts.

(In theory, we should be able to pick any value, like 0, and use it consistently in the export; the import tool already supports renumbering a realm).